test/bit_stream.cpp

```diff
@@ -0,0 +1,1026 @@
+using spvutils::StreamToBitset;
+using spvutils::BitsetToStream;
+using spvutils::BitsToStream;
+using spvutils::StreamToBits;
+using spvutils::GetLowerBits;
+using spvutils::EncodeZigZag;
+using spvutils::DecodeZigZag;
+
+// A simple and inefficient implementatition of BitWriterInterface,
+// using std::stringstream. Intended for tests only.
+class BitWriterStringStream : public BitWriterInterface {
+ public:
+  void WriteStream(const std::string& bits) override {
+    ss_ << bits;
+  }
+
+  void WriteBits(uint64_t bits, size_t num_bits) override {
+    assert(num_bits <= 64);
+    ss_ << BitsToStream(bits, num_bits);
+  }
+
+  size_t GetNumBits() const override {
+    return ss_.str().size();
+  }
+
+  std::vector<uint8_t> GetDataCopy() const override {
+    return StreamToBuffer<uint8_t>(ss_.str());
+  }
+
+  std::string GetStreamRaw() const {
+    return ss_.str();
+  }
+
+ private:
+  std::stringstream ss_;
+};
+
+// A simple and inefficient implementatition of BitReaderInterface.
+// Intended for tests only.
+class BitReaderFromString : public BitReaderInterface {
+ public:
+  explicit BitReaderFromString(std::string&& str)
+      : str_(std::move(str)), pos_(0) {}
+
+  explicit BitReaderFromString(const std::vector<uint64_t>& buffer)
+      : str_(BufferToStream(buffer)), pos_(0) {}
+
+  explicit BitReaderFromString(const std::vector<uint8_t>& buffer)
+      : str_(PadToWord<64>(BufferToStream(buffer))), pos_(0) {}
+
+  size_t ReadBits(uint64_t* bits, size_t num_bits) override {
+    if (ReachedEnd())
+      return 0;
+    std::string sub = str_.substr(pos_, num_bits);
+    *bits = StreamToBits(sub);
+    pos_ += sub.length();
+    return sub.length();
+  }
+
+  bool ReachedEnd() const override {
+    return pos_ >= str_.length();
+  }
+
+  const std::string& GetStreamPadded64() const {
+    return str_;
+  }
+
+ private:
+  std::string str_;
+  size_t pos_;
+};
+
+TEST(ZigZagCoding, Encode) {
+  EXPECT_EQ(0, EncodeZigZag(0));
+  EXPECT_EQ(1, EncodeZigZag(-1));
+  EXPECT_EQ(2, EncodeZigZag(1));
+  EXPECT_EQ(3, EncodeZigZag(-2));
+  EXPECT_EQ(4, EncodeZigZag(2));
+  EXPECT_EQ(5, EncodeZigZag(-3));
+  EXPECT_EQ(6, EncodeZigZag(3));
+}
+
+TEST(ZigZagCoding, Decode) {
+  EXPECT_EQ(0, DecodeZigZag(0));
+  EXPECT_EQ(-1, DecodeZigZag(1));
+  EXPECT_EQ(1, DecodeZigZag(2));
+  EXPECT_EQ(-2, DecodeZigZag(3));
+  EXPECT_EQ(2, DecodeZigZag(4));
+  EXPECT_EQ(-3, DecodeZigZag(5));
+  EXPECT_EQ(3, DecodeZigZag(6));
+}
+
+TEST(ZigZagCoding, Encode0) {
+  EXPECT_EQ(0, EncodeZigZag(0, 0));
+  EXPECT_EQ(1, EncodeZigZag(-1, 0));
+  EXPECT_EQ(2, EncodeZigZag(1, 0));
+  EXPECT_EQ(3, EncodeZigZag(-2, 0));
+}
+
+TEST(ZigZagCoding, Decode0) {
+  EXPECT_EQ(0, DecodeZigZag(0, 0));
+  EXPECT_EQ(-1, DecodeZigZag(1, 0));
+  EXPECT_EQ(1, DecodeZigZag(2, 0));
+  EXPECT_EQ(-2, DecodeZigZag(3, 0));
+}
+
+TEST(ZigZagCoding, Decode0SameAsNormalZigZag) {
+  for (int32_t i = -10000; i < 10000; i += 123) {
+    ASSERT_EQ(DecodeZigZag(i), DecodeZigZag(i, 0));
```

Comment:
It's intentional because of the loop. Also fixed in a few other places, where I used EXPECT inside a big loop.